File: topics/01-mini-grep/cpp/src/main.cpp
#include "topic.hpp"
#include <iostream>
#include <stdexcept>
#include <string>

namespace {
CliArgs parse_args(int argc, char** argv) {
  if (argc - 1 < 2 || argc - 1 > 2) {
    throw std::runtime_error("usage: t01_mini_grep_cpp <input> <query>");
  }
  CliArgs parsed{};
  parsed.input = argv[1];
  parsed.query = argv[2];
  return parsed;
}
}

int main(int argc, char** argv) {
  try {
    auto args = parse_args(argc, argv);
    std::cout << run(args);
    return 0;
  } catch (const std::exception& ex) {
    std::cerr << ex.what() << "\n";
    return 1;
  }
}

File: topics/01-mini-grep/rust/src/main.rs
fn parse_args() -> Result<t01_mini_grep_rust::CliArgs, String> {
    let argv: Vec<String> = std::env::args().skip(1).collect();
    if argv.len() < 2 || argv.len() > 2 {
        return Err("usage: t01_mini_grep_rust <input> <query>".to_string());
    }
    let input = argv[0].clone();
    let query = argv[1].clone();
    Ok(t01_mini_grep_rust::CliArgs {
        input,
        query,
    })
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = parse_args().map_err(|message| std::io::Error::new(std::io::ErrorKind::InvalidInput, message))?;
    print!("{}", t01_mini_grep_rust::run(&args));
    Ok(())
}
